src/codegen/codegen.utils.ts

```diff
@@ -19,3 +22,65 @@ export function stringify(value: unknown): string {
 
   return `${value}`
 }
+
+function isRedirect(response: ProxyData['response']) {
+  if (!response) return false
+  return [301, 302].includes(response.statusCode)
+}
+
+function getRedirectLocation(response: ProxyData['response']) {
+  if (!response || !isRedirect(response)) return undefined
+  return getLocationHeader(response.headers)
+}
+
+export function mergeRedirects(recording: ProxyData[]) {
+  const result = []
+  const processed = new Set()
+
+  for (const item of recording) {
+    const { response, id } = item
+
+    // Skip requests that have been processed
+    if (processed.has(id)) {
+      continue
+    }
+
+    // Requests without response don't need to be merged
+    if (!response) {
+      result.push(item)
+      processed.add(id)
+      continue
+    }
+
+    // Requests that are not redirects don't need to be merged
+    if (!isRedirect(response)) {
+      result.push(item)
+      processed.add(id)
+      continue
+    }
+
+    let finalResponse: ProxyData['response'] = response
+    let nextUrl = getRedirectLocation(response)
+
+    // Find the final response by following the redirect chain
+    while (nextUrl) {
+      const nextRequest = recording.find((req) => req.request.url === nextUrl)
+      if (nextRequest) {
+        processed.add(nextRequest.id)
+        finalResponse = nextRequest.response
+        nextUrl = getRedirectLocation(finalResponse)
+      } else {
+        break
+      }
+    }
+
+    result.push({
+      ...item,
+      response: finalResponse,
+    })
+
+    processed.add(id)
+  }
+
+  return result
+}
```

Comment:
Here's a simplified version, converted to functional programming approach. 
It resolves @Llandy3d's issue as well: the Location header in har file doesn't include host (e.g. "/") so next request's URL doesn't match. This solution skips URL checking. 

**One gotcha:**
This solution assumes all redirect chain's requests are in consecutive order. I wasn't to break this during my testing, but in theory could be possible in this scenario:

1. User clicks a link and a request is sent
2. A background fetch is triggered in JS to track user's click on a link
3. Response to 1. request is received, which is a redirect. 

In that case, the redirect chain would be interrupted by request in step no. 2.

To address this we would need to scan all requests and find the one matching the redirect, no longer assuming they are in order, and convert redirect header URLs to absolute when comparing. Though I'm not sure if that edge case needs to be covered at the current stage, WDYT @Llandy3d @cristianoventura ?

```suggestion
export function mergeRedirects(recording: ProxyData[]) {
  return recording.reduce((acc, request) => {
    const previousRequest = acc[acc.length - 1]

    // If the previous request is a redirect copy the response from the current
    // request to the previous request
    if (previousRequest && isRedirect(previousRequest?.response)) {
      return [
        ...acc.slice(0, -1),
        {
          ...previousRequest,
          response: request.response,
        },
      ]
    }

    return [...acc, request]
  }, [] as ProxyData[])
}

```

Reply:
@e-fisher I like how simple your solution is! 🤩 My initial implementation of using a `Set` was to keep track of what URLs were visited so the redirect chain could be resolved regardless of the order. We may run into issues if we assume they're always in order (which may not always be the case as you pointed out).

> and convert redirect header URLs to absolute when comparing
The current comparison uses the absolute URL when trying to find the next request, but this isn't always the case, which is why it's failing for the recording @Llandy3d  submitted.

```js
const nextRequest = recording.find((req) => req.request.url === nextUrl)
```

I'll implement a fix for this!

Reply:
I think it might make sense to try to dedicate some more time now to harden the checking so that we can also handle out of order redirects.
It's uncommon now for our tests to find that scenario but as this gets more adoption the probabilities of hitting it become higher 🤔